cloudbuild/python-scripts/utils/utils.py

```diff
@@ -0,0 +1,184 @@
+        'large_row_table', 'complex_schema_table', 'unbounded_table'
+      schema: the table schema to be inserted.
+      number_of_rows_per_batch: number of rows per thread.
+      avro_file_local: The name of the local avro file to generate the data to.
+      table_id: id of the table of the form {project_id}.{dataset_id}.{table_id}
+    """
+
+    def __init__(
+        self,
+        simple_avro_schema_string,
+        table_type,
+        avro_file_local,
+        number_of_rows_per_batch,
+        table_id,
+    ):
+        """Constructor definition for the class.
+
+        Args:
+          simple_avro_schema_string:
+          table_type: tableType Object. For custom schema and entry generation.
+          avro_file_local: The name of the local avro file to generate the data to.
+          number_of_rows_per_batch: number of rows per thread.
+          table_id: id of the table of the form {project_id}.{dataset_id}.{table_id}
+        """
+        self.table_type = table_type
+        self.schema = avro.schema.parse(simple_avro_schema_string)
+        self.number_of_rows_per_batch = number_of_rows_per_batch
+        self.avro_file_local = avro_file_local
+        self.table_id = table_id
+
+    def write_avros(self, thread_number, partition_number, current_timestamp):
+        """Method to generate fake records for BQ table.
+
+        Args:
+          thread_number: The thread number - helps in parallelism
+          partition_number: The partition number being created - only relevant in
+            partitioned table creation.
+          current_timestamp: timestamp, one hour within which timestamp entries need
+            to be generated.
+
+        Raises:
+          RuntimeError: when invalid table_type is provided.
+        """
+
+        writer = avro.datafile.DataFileWriter(
+            open(
+                self.avro_file_local.replace('.', '_' + thread_number + '.'), 'wb'
+            ),
+            avro.io.DatumWriter(),
+            self.schema,
+        )
+        self.table_type.write_rows(
+            self.number_of_rows_per_batch,
+            writer,
+            partition_number,
+            current_timestamp,
+        )
+        writer.close()
+
+    def transfer_avro_to_bq_table(self, thread_number):
+        """Method to load the created rows to BQ.
+
+        Args:
+          thread_number: The number of threads to perform the function concurrently
+            to add the avro rows to.
+        """
+        client = bigquery.Client()
+
+        job_config = bigquery.LoadJobConfig(
+            create_disposition=bigquery.CreateDisposition.CREATE_IF_NEEDED,
+            write_disposition=bigquery.WriteDisposition.WRITE_APPEND,
+            source_format=bigquery.SourceFormat.AVRO,
+            use_avro_logical_types=True,
+        )
+
+        local_avro_file = self.avro_file_local.replace(
+            '.', '_' + thread_number + '.'
+        )
```

Comment:
What exactly is being done here?

Reply:
Avro files have generic names e.g. "filename.avro". But, we write and upload several
avro files concurrently, to prevent race conditions we write and read via separate
files having names according to the thread numbers. "filename.avro" is changed to
"filename_<thread_number>.avro"

Reply:
How do you assign and keep track of "thread numbers"?
Also, as I understand it, `thread number` is an identifier/suffix and should be named as such. This method (or file in general) does not need to care how the identifier/suffix is derived. Please consider renaming to `filename_identifier` or `filename_suffix`.

Reply:
The user/we (since this is for testing) can decide the number of threads we wanna deploy.
thread number is simply obtained from the iterator in a loop where we create the threads. In other words thread_number is simply an integer in [0,1,..., n-1] where n is the total number of threads we/user decides to use.

Reply:
Fixed now.